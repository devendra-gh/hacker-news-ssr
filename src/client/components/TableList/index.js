/* eslint-disable jsx-a11y/no-interactive-element-to-noninteractive-role */
import React from 'react';
import PropTypes from 'prop-types';
import ReactTimeAgo from 'react-time-ago';
import JavascriptTimeAgo from 'javascript-time-ago';
import en from 'javascript-time-ago/locale/en';
import { upvote, tableLoader } from '../../utils/style';

JavascriptTimeAgo.addLocale(en);

const TableList = (props) => {
  const { newsList, isLoading, onClickNewsList } = props;

  return (
    <table className="striped responsive-table" role="contentinfo" aria-label="Hacker News List">
      <thead>
        <tr>
          <th>Comments</th>
          <th>Vote count</th>
          <th>UpVote</th>
          <th>New Details</th>
          <th>&nbsp;</th>
        </tr>
      </thead>
      <tbody>
        {newsList && newsList.length ? (
          newsList.map((news) => {
            return (
              <tr key={`tr-${news.objectID}-key`} title={news.objectID}>
                <td>{news.num_comments || 0}</td>
                <td>{news.points || 0}</td>
                <td>
                  <button
                    role="navigation"
                    aria-label="UP Vote"
                    disabled={isLoading}
                    onClick={() => {
                      onClickNewsList({
                        objectID: news.objectID,
                        hide: false,
                        points: news.points + 1,
                      });
                    }}
                    type="button"
                    style={upvote(isLoading)}
                  >
                    <i className="material-icons medium">arrow_drop_up</i>
                  </button>
                </td>
                <td>
                  <h6>{`${news.title}`}</h6>
                  <small>{`(${news.url}) `}</small>
                  <small>by</small>
                  <strong>{` ${news.author} `}</strong>
                  <small>
                    {news.created_at instanceof Date && !isNaN(news.created_at) ? (
                      <ReactTimeAgo
                        title={new Date(news.created_at)}
                        date={new Date(news.created_at)}
                      />
                    ) : (
                      news.created_at
                    )}
                  </small>
                </td>
                <td>
                  <button
                    role="navigation"
                    aria-label="Hide News"
                    className="btn-small"
                    type="button"
                    onClick={() => {
                      onClickNewsList({
                        objectID: news.objectID,
                        hide: true,
                        points: news.points,
                      });
                    }}
                  >
                    hide
                  </button>
                </td>
              </tr>
            );
          })
        ) : (
          <tr>
            <td colSpan="5">
              {isLoading ? (
                <div style={tableLoader}>
                  <div className="preloader-wrapper active">
                    <div className="spinner-layer spinner-red-only">
                      <div className="circle-clipper left">
                        <div className="circle" />
                      </div>
                      <div className="gap-patch">
                        <div className="circle" />
                      </div>
                      <div className="circle-clipper right">
                        <div className="circle" />
                      </div>
                    </div>
                  </div>
                </div>
              ) : (
                <div>No Data Found</div>
              )}
            </td>
          </tr>
        )}
      </tbody>
    </table>
  );
};

TableList.propTypes = {
  newsList: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.string,
      num_comments: PropTypes.number,
      points: PropTypes.number,
      title: PropTypes.string,
      url: PropTypes.string,
      author: PropTypes.string,
      created_at: PropTypes.date,
    })
  ),
  isLoading: PropTypes.bool,
  onClickNewsList: PropTypes.func,
};

TableList.defaultProps = {
  newsList: {
    objectID: '',
    num_comments: 0,
    points: 0,
    title: '',
    url: '',
    author: '',
    created_at: '',
  },
  isLoading: false,
  onClickNewsList: null,
};

export default TableList;
