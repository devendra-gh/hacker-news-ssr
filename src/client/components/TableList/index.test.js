import React from 'react';
import { render, cleanup, waitFor, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import TableList from './index';

describe('<TableList />', () => {
  let props;

  beforeEach(() => {
    props = {
      newsList: [
        {
          objectID: '12345',
          num_comments: 10,
          points: 100,
          title: 'Test',
          url: 'https://www.google.com/search?q=test',
          author: 'Test',
          created_at: new Date('Wed Jul 30 2020 23:41:03 GMT+0400 (Gulf Standard Time)'),
        },
      ],
      isLoading: false,
      onClickNewsList: jest.fn(),
    };
  });

  afterEach(cleanup);

  test('Renders <TableList /> component correctly', async () => {
    const { container, getByText } = render(
      <TableList
        newsList={props.newsList}
        isLoading={props.isLoading}
        onClickNewsList={props.onClickNewsList}
      />
    );

    const upVoteButton = await waitFor(() => getByText('arrow_drop_up'));
    const hideButton = await waitFor(() => getByText('hide'));
    expect(container).toMatchSnapshot();
    fireEvent.click(upVoteButton);
    fireEvent.click(hideButton);
  });
});
