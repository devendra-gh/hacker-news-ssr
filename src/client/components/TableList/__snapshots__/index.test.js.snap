// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableList /> Renders <TableList /> component correctly 1`] = `
<div>
  <table
    aria-label="Hacker News List"
    class="striped responsive-table"
    role="contentinfo"
  >
    <thead>
      <tr>
        <th>
          Comments
        </th>
        <th>
          Vote count
        </th>
        <th>
          UpVote
        </th>
        <th>
          New Details
        </th>
        <th>
          Â 
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        title="12345"
      >
        <td>
          10
        </td>
        <td>
          100
        </td>
        <td>
          <button
            aria-label="UP Vote"
            role="navigation"
            style="padding: 0px; font-size: 0px; width: 60px; height: 60px; background-color: rgba(0, 0, 0, 0.05); cursor: pointer; overflow: hidden;"
            type="button"
          >
            <i
              class="material-icons medium"
            >
              arrow_drop_up
            </i>
          </button>
        </td>
        <td>
          <h6>
            Test
          </h6>
          <small>
             https://www.google.com/search?q=test 
          </small>
          <small>
            by
          </small>
          <strong>
             Test 
          </strong>
          <small>
            <time
              datetime="2020-07-30T19:41:03.000Z"
              title="Thu Jul 30 2020 23:41:03 GMT+0400 (Gulf Standard Time)"
            >
              in 3 hours
            </time>
          </small>
        </td>
        <td>
          <button
            aria-label="Hide News"
            class="btn-small"
            role="navigation"
            type="button"
          >
            hide
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
